### PR TITLE
[GPU] WA: Fix attention sink handling in FlashAttnV2 PA SDPA path

### PR DESCRIPTION
**Pull request by wilson-seok**
### Details:
- On NVL-P (xe3p), the PA SDPA micro-kernel fails to compile for head_size=64 (no matching kernel in xe3_h128 config), falling back to the FlashAttnV2 online-softmax path in sdpa_opt.cl. This path was missing the attention sink correction, causing significant accuracy degradation (~80%) for models using attention sinks (e.g., sliding window + sink tokens).

Fix:
- This is WA code for NVL-P. But fix in sdpa_opt is required. I will remove condition of fallback to sdpa opt when NVL-P has proper sdpa micro kernel config.
- paged_attention_opt.cpp: Bind sink buffer and define HAS_SINK_INPUT JIT for the PagedAttentionSDPAOptGeneratorMultiToken path
- sdpa_opt.cl: Add post-partition sink correction in the FlashAttnV2 IS_FLASHATTEN_V2 path - after all KV partitions are processed, incorporate the sink logit into the softmax denominator rescaling

The sink mechanism injects a per-head scalar logit into softmax without a corresponding V contribution, acting as a dampening factor. Without this correction, full probability mass goes to real KV tokens, inflating the attention output.

### Tickets:
 - 186515

### AI Assistance:
 - *AI assistance used: yes*
 - *If yes, summarize how AI was used and what human validation was performed (build/tests/manual checks).*
 Analyzed with AI support, validation on cross platform by human.

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl [hunk 1]
```diff
@@ -2546,6 +2549,26 @@ KERNEL(sdpa_opt)(
         } /* end of QK*V calculation */
     } /* end of iter over source sequence length */
 
+#if IS_FLASHATTEN_V2 && defined(HAS_SINK_INPUT)
```
Comment (michal-miotk) [inline src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl]: HAS_SINK_INPUT will be better if one day somebody define it with 0

Comment (michal-miotk) [reply src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl]: I mean defined(HAS_SINK_INPUT) -> HAS_SINK_INPUT

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl]: Updated! Thanks!

File: src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl [hunk 2]
```diff
@@ -2546,6 +2549,26 @@ KERNEL(sdpa_opt)(
...
+    // Apply attention sink after all KV partitions are processed.
+    // Sink adds a virtual logit to the softmax denominator with zero V contribution.
+    {
+        SOFTMAX_ACCUMULATOR_TYPE sink_val = TO_SOFTMAX_ACCUMULATOR_TYPE(sink_ptr[b1_idx]);
```
[the remark this answers is not in the document]
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl]: Added test case!

File: src/plugins/intel_gpu/tests/unit/test_cases/paged_attention_gpu_test.cpp [hunk 3]
```diff
@@ -2972,6 +2972,537 @@ static std::vector<int32_t> gen_tokens_ids_test_data(size_t seq_len, int num_ima
...
+        auto pa_prim = paged_attention("paged_attention", pa_inputs);
+        pa_prim.k_head_size = p.k_head_size;
+        pa_prim.v_head_size = p.v_head_size;
+        pa_prim.kv_heads_num = p.num_kv_heads;
+        pa_prim.heads_num = p.num_heads;
+        pa_prim.scale_val = pam.get_default_scale();
+        pa_prim.has_alibi = false;
+        pa_prim.has_token_type_ids = false;
+        pa_prim.has_sink_input = true;
+        pa_prim.num_outputs = 1;
```
[the remark this answers is not in the document]
Comment (wilson-seok) [reply src/plugins/intel_gpu/tests/unit/test_cases/paged_attention_gpu_test.cpp]: Updated!

**Review by hyunback — APPROVED** [review]
LGTM. But there are too many duplicated in unit-test code, Simply could we use the existing execute function a bit more? (Frankly it seems need to refactoring..)

**Comment by wilson-seok** [comment]
> LGTM. But there are too many duplicated in unit-test code, Simply could we use the existing execute function a bit more? (Frankly it seems need to refactoring..)

I did code refactoring in the test code. Thanks!

File: src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl [hunk 4]
```diff
@@ -2546,6 +2549,26 @@ KERNEL(sdpa_opt)(
...
+            SOFTMAX_ACCUMULATOR_TYPE correction = native_exp(max_prev - max_new);
+            // Rescale output_acc (all work items do this for their own register)
+            output_acc[seq_idx] = TO_OUTPUT_TYPE(TO_SOFTMAX_ACCUMULATOR_TYPE(output_acc[seq_idx]) * correction);
+            // Update exp_sum (only one thread per seq_idx)
+            if (sgid == 0 && sglid == 0) {
+                slm_exp_sum_prev[seq_idx] = slm_exp_sum_prev[seq_idx] * correction + native_exp(sink_val - max_new);
+            }
+        }
+    }
+    barrier(CLK_LOCAL_MEM_FENCE);
```
Comment (e-ddykim) [inline src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl]: Do we need `barrier` here?

File: src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl [hunk 5]
```diff
@@ -2546,6 +2549,26 @@ KERNEL(sdpa_opt)(
         } /* end of QK*V calculation */
     } /* end of iter over source sequence length */
 
+#if IS_FLASHATTEN_V2 && HAS_SINK_INPUT
```
Comment (e-ddykim) [inline src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl]: Why do we need to check `IS_FLASHATTEN_V2`?

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl]: - The non-V2 path normalizes output_acc inline within each partition iteration by dividing by updated_total_exp_sum. By the time the loop finishes, output_acc is already fully normalized.
- The V2 path accumulates un-normalized output, only dividing by slm_exp_sum_prev at the final output stage (output_acc[seq_idx] /= slm_exp_sum_prev[seq_idx]).
- This sink correction modifies slm_exp_sum_prev and rescales output_acc — which only makes sense in the V2 data flow. Applying it to the non-V2 path would be incorrect (double-normalization).
- Additionally, slm_max_val_prev/slm_exp_sum_prev are only maintained correctly through the V2 loop path.

File: src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl (hunk 4 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl]: Removed it. Thanks!

File: src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl (hunk 5 above)
Comment (e-ddykim) [reply src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl]: `SINK_INPUT` should likely be applied independently of `IS_FLASHATTEN_V2`. It's unclear why its behavior is conditioned on `IS_FLASHATTEN_V2`.

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/graph/impls/ocl_v2/sdpa_opt.cl]: You are right. I thought non-V2 path would have their own calculation and it was wrong.
I removed IS_FLASHATTEN_V2 here. Thanks!

File: src/plugins/intel_gpu/tests/unit/test_cases/paged_attention_gpu_test.cpp [hunk 6]
```diff
@@ -1804,13 +1804,23 @@ struct PagedAttentionTest : public ::testing::TestWithParam<T> {
     cldnn::engine& engine = get_test_engine();
     float tolerance = 2e-3;
     memory::ptr last_key_cache_mem = nullptr;
+    memory::ptr last_output_data_mem = nullptr;
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/unit/test_cases/paged_attention_gpu_test.cpp:1807]: nit: what about `output_data_mem` instead of `last_output_data_mem`?

File: src/plugins/intel_gpu/tests/unit/test_cases/paged_attention_gpu_test.cpp [hunk 7]
```diff
@@ -2972,6 +2997,60 @@ static std::vector<int32_t> gen_tokens_ids_test_data(size_t seq_len, int num_ima
     return v;
 }
 
+
+// Test class to verify FlashAttnV2 multi-token sink correction is actually active.
+// Runs the same prompt twice with FA_V2 enabled: once with non-zero sinks, once with zero sinks.
+// Asserts outputs differ, proving the sink code at line 2552 in sdpa_opt.cl executes.
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/unit/test_cases/paged_attention_gpu_test.cpp]: do not mention line number in the test code because it is not possible to maintain.

**Review by isanghao — APPROVED** [review]

File: src/plugins/intel_gpu/tests/unit/test_cases/paged_attention_gpu_test.cpp (hunk 6 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/tests/unit/test_cases/paged_attention_gpu_test.cpp:1807]: The last_ prefix is intentional and consistent with other members in the same class. And it is actually last execution result. And output_data_mem already exists in below line. So I will keep this name.

File: src/plugins/intel_gpu/tests/unit/test_cases/paged_attention_gpu_test.cpp (hunk 7 above)
Comment (wilson-seok) [reply src/plugins/intel_gpu/tests/unit/test_cases/paged_attention_gpu_test.cpp]: Fixed it. Thanks!

**Comment by wilson-seok** [comment]
No regression from daily
